test/functional/url-expander/test-expander.js

```diff
@@ -140,108 +140,248 @@ describes.realWin('Expander', {
+        expect(expander.expand('CONCAT(aaa, UPPERCASE(bbb))', mockBindings))
+            .to.eventually.equal('aaaBBB')
+      );
+
+      it('parses function with two funcs as args', () => {
+        const url = 'CONCAT(LOWERCASE(AAA),UPPERCASE(bbb)';
+        return expect(expander.expand(url, mockBindings))
+            .to.eventually.equal('aaaBBB');
+      });
+
+      it('parses function with three funcs as args', () => {
+        const url = 'CAT_THREE(LOWERCASE(AAA),UPPERCASE(bbb),LOWERCASE(CCC))';
+        return expect(expander.expand(url, mockBindings))
+            .to.eventually.equal('aaaBBBccc');
+      });
+
+      it('should handle real urls', () => {
+        const url = 'http://www.amp.google.com/?client=CLIENT_ID(__ga)&canon=CANONICAL_URL&random=RANDOM';
+        const expected = 'http://www.amp.google.com/?client=amp-GA12345&canon=www.google.com&random=123456';
+        return expect(expander.expand(url, mockBindings))
+            .to.eventually.equal(expected);
+      });
+
+      it('should treat unrecognized keywords as normal strings', () => {
+        return expect(expander.expand('TRIM(FAKE(aaaaa))', mockBindings))
+            .to.eventually.equal('');
+      });
+
+      it('ignores commas within backticks', () => {
+        const url = 'CONCAT(`he,llo`,UPPERCASE(world)';
+        return expect(expander.expand(url, mockBindings))
+            .to.eventually.equal('he,lloWORLD');
+      });
+
+      it('ignores left parentheses within backticks', () => {
+        const url = 'CONCAT(hello, `wo((rld`)';
+        return expect(expander.expand(url, mockBindings))
+            .to.eventually.equal('hellowo((rld');
+      });
+
+      it('ignores right parentheses within backticks', () => {
+        const url = 'CONCAT(`hello)`,UPPERCASE(world)';
+        return expect(expander.expand(url, mockBindings))
+            .to.eventually.equal('hello)WORLD');
+      });
+
+      it('throws on bad input with back ticks', () => {
+        const url = 'CONCAT(bad`hello`, world)';
+        allowConsoleError(() => { expect(() => {
+          expander.expand(url, mockBindings);
+        }).to.throw(/bad/); });
+      });
+
+      it('should handle tokens with parenthesis next to each other', () => {
+        const url = 'http://www.google.com/?test=RANDOMCLIENT_ID(__ga)UPPERCASE(foo)';
+        const expected = 'http://www.google.com/?test=123456amp-GA12345FOO';
+        return expect(expander.expand(url, mockBindings))
+            .to.eventually.equal(expected);
+      });
     });
 
-    it('should handle tokens with parenthesis next to each other', () => {
-      const url = 'http://www.google.com/?test=RANDOMCLIENT_ID(__ga)UPPERCASE(foo)';
-      const expected = 'http://www.google.com/?test=123456amp-GA12345FOO';
-      return expect(expander.expand(url, mockBindings))
-          .to.eventually.equal(expected);
+    describe('#expand sync', () => {
+
+      it('should handle empty urls', () =>
```

Comment:
seems we are duplicating most of the test cases.
can you share the tests?

Reply:
maybe it's easier to do things like:

```
const testCases = [
  {
     description: 'parses one function, one argument'
     input: 'TRIM(aaaa    )',
     output: 'aaaa',
  }, {
     ...
  }
];

testCase.forEach(...)
```

Reply:
refactored.